src/amqp/connection.js

```diff
@@ -195,8 +193,8 @@ Adapter.prototype.bumpIndex = function () {
 Adapter.prototype.getNextUri = function () {
   const server = this.getNext(this.servers);
   const port = this.getNext(this.ports);
-  const uri = getUri(this.protocol, this.user, escape(this.pass), server, port, this.vhost, this.heartbeat);
-  return uri;
+  const uri = getUri(this.protocol, encodeURIComponent(this.user), encodeURIComponent(this.pass), server, port, this.vhost, this.heartbeat);
+  return [uri, server];
```

Comment:
A couple additional benefits of using URL as mentioned in the other comment around `getUri()` would be that the encoding of user|password are taken care of for us and if a proper url cannot be constructed the error will be obvious prior to even attempting to connect.

Reply:
Thanks for the response. 

I see what you mean and agree with the sentiment of
> Either we encode the string based on an object, or the end user does when they give a URI

(And in case anyone else comes along and is curious https://github.com/onebeyond/rascal/blob/0a58d1c12534b8ce8a9067e01ff0e5f41aa0dacb/lib/config/configure.js#L127 is where the uri is built in rascal.)

I think we are still at risk for double encoding with this change though because of the 

```
const uriOpts = parseUri(parameters.uri);
Object.assign(parameters, uriOpts);
```

lines here https://github.com/Foo-Foo-MQ/foo-foo-mq/pull/56/files#diff-7a2ac2ec921e0936835d89c9e72c4aa24be572652934b2472571767e4096ba01R97-R98

Rascal does a decode on the username and password here https://github.com/onebeyond/rascal/blob/0a58d1c12534b8ce8a9067e01ff0e5f41aa0dacb/lib/config/configure.js#L94 and in our scenario we currently are not decoding. 

So using `encodeURIComponent` when the the `uri` connection string is provided in the options https://github.com/Foo-Foo-MQ/foo-foo-mq/blob/main/docs/connections.md#rabbitaddconnection--options- I believe will result in the double encode scenario as well wouldn't it? 

Maybe we also need the decodeURIComponent here https://github.com/Foo-Foo-MQ/foo-foo-mq/pull/56/files#diff-7a2ac2ec921e0936835d89c9e72c4aa24be572652934b2472571767e4096ba01R65-R66

Reply:
> > Either we encode the string based on an object, or the end user does when they give a URI
> 
> (And in case anyone else comes along and is curious https://github.com/onebeyond/rascal/blob/0a58d1c12534b8ce8a9067e01ff0e5f41aa0dacb/lib/config/configure.js#L127 is where the uri is built in rascal.)
> 
> I think we are still at risk for double encoding with this change though because of the
> 
> ```
> const uriOpts = parseUri(parameters.uri);
> Object.assign(parameters, uriOpts);
> ```
> 
> lines here https://github.com/Foo-Foo-MQ/foo-foo-mq/pull/56/files#diff-7a2ac2ec921e0936835d89c9e72c4aa24be572652934b2472571767e4096ba01R97-R98

I admit this one slipped my mind, I think a simple solution would be to follow Rascals example and decode. We could indeed decode. It is an easy change: something like this:

```js
function parseUri (uri) {
  if (uri) {
    const parsed = new url.URL(uri);
    const heartbeat = parsed.searchParams.get('heartbeat');
    return {
      useSSL: parsed.protocol.startsWith('amqps'),
      user: decodeURIComponent(parsed.username),
      pass: decodeURIComponent(parsed.password),
      host: parsed.hostname,
      port: parsed.port,
      vhost: parsed.pathname ? parsed.pathname.slice(1) : undefined,
      heartbeat: heartbeat
    };
  }
}
```

Reply:
I pushed the change.